File: cpp/strings/0038-count-and-say.cpp
#include <algorithm>
#include <string>

using namespace std;


string countAndSay(int n) {
    string response = "";
    if (n == 1) return "1";
    if (n == 2) return "11";
    
    string say = countAndSay(n - 1);
    int i = 0;
    
    while (i < say.length()) {
        int count  = 0;
        char cur = say.at(i);
        while(i < say.length() && say.at(i) == cur) {
            count++;
            i++;
        }
        
        response.push_back(count + '0');
        response.push_back(cur);
    }
    
    return response;
}

File: js/strings/0038-count-and-say.js
var countAndSay = function(n) {
    if (n == 1) return "1";
    if (n == 2) return "11"
    
    var say = countAndSay(n-1)
    return count(say);
};

var count = function(str) {
    var response = "";
    var cur = str[0];
    var counter = 1;
    
    for (var i = 1; i < str.length; i++) {
        if (str[i] == cur) counter++;
        else {
            response += counter + cur;
            counter = 1;
            cur = str[i];
        }
    }
    response += counter + cur;
    
    return response;
};